File: swift/InkRecognition/Recognizer/Recognizer/InkRoot.swift

import Foundation

@objc
enum RecognitionResultStatus : Int {
    case unchanged,
    updated,
    failed
}

@objc
class InkRecognitionError : NSObject {
    var code : String = ""
    var message : String  = ""
    var target : String = ""
    var details = [InkRecognitionError]()
    
    @objc
    init(jsonString: [String: Any], isAccessError : Bool) {
        if !isAccessError {
            code = jsonString["code"] as? String ?? ""
            message = jsonString["message"] as? String ?? ""
            target = jsonString["target"] as? String ?? ""
            if let detailsList = jsonString["details"] as? [[String: Any]] {
                for detailUnit in detailsList {
                    details.append(InkRecognitionError(jsonString: detailUnit, isAccessError: false))
                }
            }
        } else {
            //Access errors have a slightly different JSON format
            code = "AccessDenied"
            if let errorDetail = jsonString["error"] as? [String: Any] {
                message = errorDetail["message"] as? String ?? ""
            }
            target = "Request"
        }
    }
    
    @objc
    init(error: String)
    {
        code = "ServerError"
        message = error
        target = "Request"
    }
    
    @objc
    public func toString() -> String {
        return String("Code: \(code) \nMessage: \(message) Target: \(target) \nDetails:\n" + getErrorDetails())
    }
    
    @objc
    public func getErrorDetails() -> String {
        var errorDetails = String()
        for detailUnit in details {
            errorDetails.append(detailUnit.toString())
        }
        return errorDetails
    }
}

@objc
class InkRoot : NSObject {
    
    var resultJSON : [String: Any]
    var recognizedContainers = [Int: InkRecognitionUnit]()
    var recognizedWords = [Int: InkWord]()
    var recognizedDrawings = [Int: InkDrawing]()
    var recognitionUnits = [Int: InkRecognitionUnit]()
    var orderedWords : [String] = []
    var errorStore : InkRecognitionError!
    var resultStatus: RecognitionResultStatus = RecognitionResultStatus.unchanged
    
    @objc
    init(jsonString: [String: Any], statusCode: Int) {
        resultJSON = jsonString
        
        
        if statusCode >= 400 {
            var accessError = false
            if statusCode == 401 {
                accessError = true;
            }
            errorStore = InkRecognitionError(jsonString: resultJSON, isAccessError: accessError)
            resultStatus = RecognitionResultStatus.failed
            return
        }
        
        let jsonUnits = jsonString["recognitionUnits"] as? [[String: Any]]
        for jsonUnit in jsonUnits! {
            let category = jsonUnit["category"] as? String
            let index = jsonUnit["id"] as! Int
            switch category {
            case "inkWord":
                let inkWord = InkWord(json: jsonUnit)
                recognizedWords[index] = inkWord
                recognitionUnits[index] = inkWord
                orderedWords.append(inkWord.text)
            case "inkDrawing":
                let inkShape = InkDrawing(json: jsonUnit)
                recognizedDrawings[index] = inkShape
                recognitionUnits[index] = inkShape
            case "line":
                let inkLine = InkLine(json: jsonUnit)
                recognizedContainers[index] = inkLine
                recognitionUnits[index] = inkLine
            case "inkBullet":
                let inkBullet = InkBullet(json: jsonUnit)
                recognitionUnits[index] = inkBullet
            case "listItem":
                let inkListItem = InkListItem(json: jsonUnit)
                recognizedContainers[index] = inkListItem
                recognitionUnits[index] = inkListItem
            case "paragraph":
                let inkParagraph = InkParagraph(json: jsonUnit)
                recognizedContainers[index] = inkParagraph
                recognitionUnits[index] = inkParagraph
            case "writingRegion":
                let inkWritingRegion = InkWritingRegion(json: jsonUnit)
                recognizedContainers[index] = inkWritingRegion
                recognitionUnits[index] = inkWritingRegion
            default:
                errorStore = InkRecognitionError(error: "unknown category")
                resultStatus = RecognitionResultStatus.failed
                return
            }
        }
        resultStatus = RecognitionResultStatus.updated
    }
    
    @objc
    public func getInkWords() -> [InkWord] {
        var words = [InkWord]()
        recognizedWords.values.forEach {value in
            words.append(value)
        }
        return words
    }
    
    @objc
    public func getWords() ->[String] {
        return orderedWords
    }
    
    @objc
    public func getNodes(ids: [Int]) -> [InkRecognitionUnit] {
        var recognizedUnits = [InkRecognitionUnit]()
        for id in ids {
            recognizedUnits.append(recognitionUnits[id]!)
        }
        return recognizedUnits
    }
    
    @objc
    public func getNode(id: Int) -> InkRecognitionUnit {
        let recognizedUnit = recognitionUnits[id]!
        return recognizedUnit
    }
    
    @objc
    public func getParagraphs() -> [InkParagraph] {
        var paragraphs = [InkParagraph]()
        recognizedContainers.values.forEach {value in
            if let paragraph = value as? InkParagraph {
                paragraphs.append(paragraph)
            }
        }
        return paragraphs
    }
    
    @objc
    public func getDrawings() -> [InkDrawing] {
        var shapes = [InkDrawing]()
        recognizedDrawings.values.forEach {value in
            shapes.append(value)
        }
        return shapes
    }
}

File: java/InkRecognition/android-sample-app/Recognizer/app/src/main/java/CognitiveServices/Ink/Recognition/InkRoot.java
package CognitiveServices.Ink.Recognition;

import android.util.DisplayMetrics;

import org.json.JSONArray;
import org.json.JSONObject;

import java.net.HttpURLConnection;
import java.security.InvalidParameterException;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.Set;

class InkRoot {
    //Using different collections to isolate each type for type specific queries.
    private final Hashtable<Integer, InkRecognitionUnit> recognizedContainers = new Hashtable<>();
    private final ArrayList<InkWord> wordList = new ArrayList<>();
    private final Hashtable<Integer, InkRecognitionUnit> recognizedDrawings = new Hashtable<>();
    private final Hashtable<Integer, InkRecognitionUnit> recognizedUnits = new Hashtable<>();

    public InkRecognitionError getRecognitionError() {
        return recognitionError;
    }

    private InkRecognitionError recognitionError;

    public RecognitionResultStatus getResultStatus() {
        return resultStatus;
    }

    private RecognitionResultStatus resultStatus = RecognitionResultStatus.UNCHANGED;

    InkRoot(String json, DisplayMetrics metrics, int httpResponseCode)
    {
        try {
            JSONObject jsonResponse = new JSONObject(json);

            if (httpResponseCode == HttpURLConnection.HTTP_OK) {
                JSONArray jsonRecognitionUnits = jsonResponse.getJSONArray("recognitionUnits");
                for (int i = 0;
                     i < jsonRecognitionUnits.length() &&
                     resultStatus != RecognitionResultStatus.FAILED;
                     i++) {

                    String category = jsonRecognitionUnits.getJSONObject(i).getString("category");
                    String unitJSON = jsonRecognitionUnits.getJSONObject(i).toString();
                    int id = jsonRecognitionUnits.getJSONObject(i).getInt("id");
                    //Instantiate the recognition units
                    switch (category) {
                        case "inkWord":
                            InkWord word = new InkWord(unitJSON, metrics, this);
                            recognizedUnits.put(id, word);
                            wordList.add(word);
                            break;
                        case "inkBullet":
                            InkBullet bullet = new InkBullet(unitJSON, metrics, this);
                            recognizedUnits.put(id, bullet);
                            break;
                        case "inkDrawing":
                            InkDrawing drawing = new InkDrawing(unitJSON, metrics, this);
                            recognizedDrawings.put(id, drawing);
                            recognizedUnits.put(id, drawing);
                            break;
                        case "line":
                            InkLine line = new InkLine(unitJSON, metrics, this);
                            recognizedContainers.put(id, line);
                            recognizedUnits.put(id, line);
                            break;
                        case "listItem":
                            InkListItem listItem = new InkListItem(unitJSON, metrics, this);
                            recognizedContainers.put(id, listItem);
                            recognizedUnits.put(id, listItem);
                            break;
                        case "paragraph":
                            InkParagraph paragraph = new InkParagraph(unitJSON, metrics, this);
                            recognizedContainers.put(id, paragraph);
                            recognizedUnits.put(id, paragraph);
                            break;
                        case "writingRegion":
                            InkWritingRegion writingRegion = new InkWritingRegion(unitJSON, metrics, this);
                            recognizedContainers.put(id, writingRegion);
                            recognizedUnits.put(id, writingRegion);
                            break;
                        default:
                            recognitionError = new InkRecognitionError("unknown unit");
                            resultStatus = RecognitionResultStatus.FAILED;
                            break;
                    }
                }
                resultStatus = RecognitionResultStatus.UPDATED;
            } else {
                //build error report
                if (httpResponseCode != 0) {
                    recognitionError = new InkRecognitionError(jsonResponse);
                } else {
                    //build default error
                    recognitionError = new InkRecognitionError();
                }
                resultStatus = RecognitionResultStatus.FAILED;
            }
        }
        catch (Exception e) {
            recognitionError = new InkRecognitionError(e.getMessage());
        }
    }

    public ArrayList<InkWord> getInkWords()
    {
        return wordList;
    }

    public ArrayList<InkParagraph> getParagraphs()  {
        ArrayList<InkParagraph> paragraphs = new ArrayList<>();
        for (Integer key : recognizedContainers.keySet()) {
            InkRecognitionUnit entryValue = recognizedContainers.get(key);
            if (entryValue.getCategory() == InkRecognitionUnitCategory.INK_PARAGRAPH) {
                paragraphs.add((InkParagraph)entryValue);
            }
        }
        return paragraphs;
    }

    public ArrayList<InkDrawing> getInkDrawings() {
        ArrayList<InkDrawing> drawings = new ArrayList<>();
        Set<Integer> keys = recognizedDrawings.keySet();
        for (Integer key: keys) {
            drawings.add((InkDrawing)recognizedDrawings.get(key));
        }
        return drawings;
    }

    public ArrayList<InkRecognitionUnit> getRecognitionUnits(ArrayList<Integer> ids) {
        ArrayList<InkRecognitionUnit> nodes = new ArrayList<>();
        for (int i=0;
             i < ids.size();
             i++) {
            InkRecognitionUnit unit = recognizedUnits.get(ids.get(i));
            if (unit != null) {
                nodes.add(unit);
            }
        }
        return nodes;
    }

    public InkRecognitionUnit getRecognitionUnit(int id) throws InvalidParameterException {
        InkRecognitionUnit node = recognizedUnits.get(id);
        if (node == null) {
            throw new InvalidParameterException("unknown id");
        }
        return node;
    }
}
